### PR TITLE
fix: [OIP-510]: Fix ELK Health source issues

### PR DESCRIPTION
**Pull request by praneshg239**
### 1. Add input + JSON Selector component
image

image

### 2. Add user to add custom value for indexes in templates
image

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by praneshg239** [comment]
retrigger titlecheck

File: src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/CommonCustomMetricFormContainer.tsx [hunk 1]
```diff
@@ -85,8 +86,16 @@ export default function CommonCustomMetricFormContainer(props: CommonCustomMetri
       setRecords([])
       setHealthSourceTimeSeriesData([])
 
+      const isELKHealthSource = sourceData?.product?.value === HealthSourceTypes.ElasticSearch_Logs
```
Comment (harshilbits) [inline src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/CommonCustomMetricFormContainer.tsx:89]: We should not write as per the ELK condition, condition should be generic for any field which is above query.

File: src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/CommonCustomMetricFormContainer.tsx [hunk 2]
```diff
@@ -85,8 +86,16 @@ export default function CommonCustomMetricFormContainer(props: CommonCustomMetri
...
+      const logIndexType = getMultiTypeFromValue(values?.index)
+
       // Fetch the records for latest query
-      if (query && !isConnectorRuntimeOrExpression && !isQueryRuntimeOrExpression) {
+      if (
+        query &&
+        !isConnectorRuntimeOrExpression &&
+        !isQueryRuntimeOrExpression &&
+        (!isELKHealthSource || logIndexType === MultiTypeInputType.FIXED)
```
Comment (harshilbits) [inline src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/CommonCustomMetricFormContainer.tsx:97]: We should not use isELK Health source condition anywhere.

File: src/modules/85-cv/components/CVMultiTypeQuery/CVMultiTypeQuery.tsx [hunk 3]
```diff
@@ -62,17 +64,19 @@ export default function CVMultiTypeQuery({
...
-        margin={{ top: 'medium' }}
-        disabled={disableFetchButton}
-        {...fetchButtonProps}
-        tooltip={runQueryBtnTooltip}
-      />
+      {!hideFetchButton && (
+        <Button
+          intent="primary"
+          text={getString('cv.monitoringSources.gcoLogs.fetchRecords')}
+          onClick={async () => {
```
Comment (harshilbits) [inline src/modules/85-cv/components/CVMultiTypeQuery/CVMultiTypeQuery.tsx:71]: Won't onClick={fetchRecords}  do ?

File: src/modules/85-cv/components/CVMultiTypeQuery/__tests__/CVMultiTypeQuery.test.tsx [hunk 4]
```diff
@@ -61,6 +61,30 @@ describe('Validate CVMultiTypeQuery', () => {
     expect(fetchRecords).toBeCalledTimes(1)
   })
 
+  test('should check fetch records button is hidden if hideFetchButton prop is true', async () => {
+    const fetchRecords = jest.fn()
+    render(
+      <TestWrapper>
+        <Formik formName="" initialValues={{ query: mockQuery }} onSubmit={() => undefined}>
+          {_ => {
```
Comment (harshilbits) [inline src/modules/85-cv/components/CVMultiTypeQuery/__tests__/CVMultiTypeQuery.test.tsx:69]: Please avoid using underscore.

File: src/modules/85-cv/components/CVMultiTypeQuery/__tests__/CVMultiTypeQuery.test.tsx [hunk 5]
```diff
@@ -61,6 +61,30 @@ describe('Validate CVMultiTypeQuery', () => {
     expect(fetchRecords).toBeCalledTimes(1)
   })
 
+  test('should check fetch records button is hidden if hideFetchButton prop is true', async () => {
+    const fetchRecords = jest.fn()
+    render(
+      <TestWrapper>
+        <Formik formName="" initialValues={{ query: mockQuery }} onSubmit={() => undefined}>
```
Comment (harshilbits) [inline src/modules/85-cv/components/CVMultiTypeQuery/__tests__/CVMultiTypeQuery.test.tsx:68]: Use onSubmit={jest.fn()}

File: src/modules/85-cv/components/CommonQueryViewer/CommonQueryViewer.tsx [hunk 6]
```diff
@@ -44,12 +45,20 @@ export function CommonQueryViewer(props: CommonQueryViewerProps): JSX.Element {
...
+  const { isTemplate, expressions, sourceData } = useContext(SetupSourceTabsContext)
   const { updateHelperContext, isQueryRuntimeOrExpression } = useCommonHealthSource()
   const isQueryButtonDisabled = useMemo(() => {
     return getIsQueryButtonDisabled({ query, loading, queryFieldIdentifier, values })
   }, [loading, query, queryFieldIdentifier, values])
 
+  const isELKHealthSource = sourceData?.product?.value === HealthSourceTypes.ElasticSearch_Logs
+
+  const hideRecords = useMemo(() => {
+    const logIndexType = getMultiTypeFromValue(values?.index)
```
Comment (harshilbits) [inline src/modules/85-cv/components/CommonQueryViewer/CommonQueryViewer.tsx:57]: No code should be hardcoded as per the index , it should read the health source config json in dynamic way.

File: src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.tsx [hunk 7]
```diff
@@ -77,8 +77,10 @@ export default function CommonCustomMetric(props: CommonCustomMetricInterface):
     const isServiceInstanceFixed = getMultiTypeFromValue(formikValues.serviceInstanceField) === MultiTypeInputType.FIXED
 
     if (isQueryRuntimeOrExpression) {
+      const isLogsHealthSource = getIsLogHealthSource(sourceData?.product?.value)
```
Comment (harshilbits) [inline src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.tsx:80]: No condition as per the logs or metric health source. Health source config json should be the source of truth.

File: src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.utils.ts [hunk 8]
```diff
@@ -149,3 +150,11 @@ export function initializeSelectedMetricsMap(
     >
   }
 }
+
+export const getIsLogHealthSource = (type: HealthSourceTypes): boolean => {
+  return [
+    HealthSourceTypes.ElasticSearch_Logs,
```
Comment (harshilbits) [inline src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.utils.ts:156]: Please no hardcoding for any specifc health sources.

Comment (praneshg239) [reply src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.utils.ts:156]: This is required as it is updating service instance field value to RUNTIME if we update Query to RUNTIME. This should not happen for logs. 

cc: @deepesh-ui

File: src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/components/CommonHealthSourceField/CommonHealthSourceField.tsx [hunk 9]
```diff
@@ -81,6 +81,16 @@ export default function CommonHealthSourceField(props: CommonHealthSourceFieldPr
                 data-testid={identifier}
                 multiTypeInputProps={{
                   expressions,
+                  /**
```
Comment (harshilbits) [inline src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/components/CommonHealthSourceField/CommonHealthSourceField.tsx:84]: Why are we ignoring this ?

File: src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/LogsTable/components/JsonSelectorWithDrawer.tsx [hunk 10]
```diff
@@ -185,15 +144,14 @@ export default function JsonSelectorWithDrawer(props: JsonSelectorWithDrawerProp
                 setMultiType={handleTemplateTypeUpdate}
               />
             ) : (
-              <JsonSelectorButton
+              <InputWithJsonSelector
```
Comment (harshilbits) [inline src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/LogsTable/components/JsonSelectorWithDrawer.tsx:147]: If we are not using the `JsonSelectorButton` component , shall we remove it ?

Comment (praneshg239) [reply src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/LogsTable/components/JsonSelectorWithDrawer.tsx:147]: yes, will delete that component

File: src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/components/CommonHealthSourceField/CommonHealthSourceField.tsx (hunk 9 above)
Comment (praneshg239) [reply src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/components/CommonHealthSourceField/CommonHealthSourceField.tsx:84]: In this we need only `allowCreatingNewItems`, as it is extending `SelectProps` it is asking for all other select props which are not needed.

File: src/modules/85-cv/components/CVMultiTypeQuery/__tests__/CVMultiTypeQuery.test.tsx (hunk 4 above)
Comment (praneshg239) [reply src/modules/85-cv/components/CVMultiTypeQuery/__tests__/CVMultiTypeQuery.test.tsx:69]: Done

File: src/modules/85-cv/components/CVMultiTypeQuery/__tests__/CVMultiTypeQuery.test.tsx (hunk 5 above)
Comment (praneshg239) [reply src/modules/85-cv/components/CVMultiTypeQuery/__tests__/CVMultiTypeQuery.test.tsx:68]: Done

File: src/modules/85-cv/components/CommonQueryViewer/CommonQueryViewer.tsx (hunk 6 above)
Comment (praneshg239) [reply src/modules/85-cv/components/CommonQueryViewer/CommonQueryViewer.tsx:57]: Done

File: src/modules/85-cv/components/CommonQueryViewer/CommonQueryViewer.tsx [hunk 11]
```diff
@@ -122,7 +132,7 @@ export function CommonQueryViewer(props: CommonQueryViewerProps): JSX.Element {
           runQueryBtnTooltip={runQueryBtnTooltip}
         />
       )}
-      {!(isQueryRuntimeOrExpression || isConnectorRuntimeOrExpression) ? (
+      {!(isQueryRuntimeOrExpression || isConnectorRuntimeOrExpression) && !hideRecords ? (
```
[the remark this answers is not in the document]
Comment (praneshg239) [reply src/modules/85-cv/components/CommonQueryViewer/CommonQueryViewer.tsx:135]: Done

File: src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.tsx (hunk 7 above)
Comment (praneshg239) [reply src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.tsx:80]: This is required as it is updating service instance field value to RUNTIME if we update Query to RUNTIME. This should not happen for logs.

cc: @deepesh-ui

File: src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/LogsTable/components/JsonSelectorWithDrawer.tsx (hunk 10 above)
Comment (praneshg239) [reply src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/LogsTable/components/JsonSelectorWithDrawer.tsx:147]: Done

File: src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/CommonCustomMetricFormContainer.tsx (hunk 1 above)
Comment (praneshg239) [reply src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/CommonCustomMetricFormContainer.tsx:89]: Done

File: src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/CommonCustomMetricFormContainer.tsx (hunk 2 above)
Comment (praneshg239) [reply src/modules/85-cv/pages/health-source/connectors/CommonHealthSource/components/CustomMetricForm/components/CommonCustomMetricFormContainer/CommonCustomMetricFormContainerLayout/CommonCustomMetricFormContainer.tsx:97]: Done

File: src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.tsx (hunk 7 above)
Comment (harshilbits) [reply src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.tsx:80]: Please read this from config itself. We can introduce a field in the health source config which will check for this but we shouldnt be adding any specific health source related code in the framework.

File: src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.utils.ts (hunk 8 above)
Comment (harshilbits) [reply src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.utils.ts:156]: same.

File: src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.tsx (hunk 7 above)
Comment (praneshg239) [reply src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.tsx]: Done

File: src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.utils.ts (hunk 8 above)
Comment (praneshg239) [reply src/modules/85-cv/pages/health-source/common/CommonCustomMetric/CommonCustomMetric.utils.ts]: Done

**Review by harshilbits — APPROVED** [review]

**Pull request merged**